File: benchmark/python/bitsbench.py
"""Bits-group benchmark (bits:: package surface).

Mirrors benchmark/mfb/src/bits.mfb test_bits_ops exactly: one coverage row that
folds a 64-bit hash with band/bor/bxor/bnot/sl/sr/sra/rl32/rr32/rl64/rr64/clz/
ctz/popCount/bswap16/bswap32/bswap64. Python ints are unbounded, so every op is
masked back to 64 bits; the accumulator is folded with XOR (never +) so the
workload never overflows. mfb Integer is signed 64-bit, so the printed checksum
is reinterpreted as a signed two's-complement value to line up with mfb.
"""
import sys

RUN = 1
now_ns = None
record = None

MASK64 = (1 << 64) - 1
MASK32 = (1 << 32) - 1
MASK16 = (1 << 16) - 1


def _band(a, b):
    return (a & b) & MASK64


def _bor(a, b):
    return (a | b) & MASK64


def _bxor(a, b):
    return (a ^ b) & MASK64


def _bnot(x):
    return (~x) & MASK64


def _sl(x, n):
    return (x << n) & MASK64


def _sr(x, n):
    # logical right shift of the 64-bit unsigned value
    return (x & MASK64) >> n


def _sra(x, n):
    # arithmetic right shift: reinterpret as signed, shift, mask back to 64 bits
    v = x & MASK64
    if v >> 63:
        v -= (1 << 64)
    return (v >> n) & MASK64


def _rl32(x, n):
    v = x & MASK32
    n &= 31
    return ((v << n) | (v >> (32 - n))) & MASK32 if n else v


def _rr32(x, n):
    v = x & MASK32
    n &= 31
    return ((v >> n) | (v << (32 - n))) & MASK32 if n else v


def _rl64(x, n):
    v = x & MASK64
    n &= 63
    return ((v << n) | (v >> (64 - n))) & MASK64 if n else v


def _rr64(x, n):
    v = x & MASK64
    n &= 63
    return ((v >> n) | (v << (64 - n))) & MASK64 if n else v


def _clz(x):
    v = x & MASK64
    if v == 0:
        return 64            # mfb clz(0) = 64
    return 63 - v.bit_length() + 1


def _ctz(x):
    v = x & MASK64
    if v == 0:
        return 64            # mfb ctz(0) = 64
    return (v & -v).bit_length() - 1


def _popcount(x):
    return bin(x & MASK64).count("1")


def _bswap16(x):
    v = x & MASK16
    return ((v & 0xFF) << 8) | ((v >> 8) & 0xFF)


def _bswap32(x):
    v = x & MASK32
    return (((v & 0x000000FF) << 24) | ((v & 0x0000FF00) << 8)
            | ((v & 0x00FF0000) >> 8) | ((v & 0xFF000000) >> 24)) & MASK32


def _bswap64(x):
    v = x & MASK64
    r = 0
    for _ in range(8):
        r = (r << 8) | (v & 0xFF)
        v >>= 8
    return r & MASK64


def _to_signed64(x):
    v = x & MASK64
    return v - (1 << 64) if v >> 63 else v


def test_bits_ops():
    times = []
    checksum = 0
    for _ in range(RUN):
        t0 = now_ns()
        acc = 0
        h = 2166136261
        for i in range(200000):
            x = i
            h = _bxor(h, _bor(x, _sl(x, 3)))
            h = _band(h, 1152921504606846975)
            h = _rl64(h, 7)
            h = _rr64(h, 3)
            h = _bxor(h, _sr(h, 11))
            h = _bxor(h, _sra(h, 2))
            h = _bxor(h, _bnot(x))
            acc = _bxor(acc, h)
            acc = _bxor(acc, _rl32(h, 5))
            acc = _bxor(acc, _rr32(h, 9))
            acc = _bxor(acc, _bswap16(x))
            acc = _bxor(acc, _bswap32(x))
            acc = _bxor(acc, _bswap64(h))
            acc = _bxor(acc, _popcount(h))
            acc = _bxor(acc, _clz(x))
            acc = _bxor(acc, _ctz(x))
        checksum = _to_signed64(acc)
        times.append(now_ns() - t0)
    print("bits_ops = %d" % checksum, file=sys.stderr)
    record("bits", "ops", times)


def run_all(run, now_ns_fn, record_fn):
    global RUN, now_ns, record
    RUN, now_ns, record = run, now_ns_fn, record_fn
    test_bits_ops()

File: benchmark/c/bitsbench.c
/* GROUP: bits (bits:: package coverage)
 *
 * Moved into its own translation unit; shared timing/recording infra comes
 * from bench.h. Mirrors bits.mfb: every bits:: member operates on raw 64-bit
 * Integer bit patterns, so a single row drives band/bor/bxor/bnot, sl/sr/sra,
 * rl32/rr32/rl64/rr64, clz/ctz/popCount and bswap16/32/64. The accumulator is
 * folded with XOR (not +) so a full-width bit pattern never trips overflow.
 *
 * All arithmetic is done on uint64_t and reported as a signed 64-bit Integer,
 * matching mfb's i64 Integer. clz(0)/ctz(0) are special-cased to 64 (mfb's
 * defined result); the 32-bit ops act on the low 32 bits and zero-extend. */
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>

#include "bench.h"
#include "bitsbench.h"

static inline uint64_t rotl64(uint64_t v, unsigned s) {
  s &= 63u;
  return s ? (v << s) | (v >> (64 - s)) : v;
}
static inline uint64_t rotr64(uint64_t v, unsigned s) {
  s &= 63u;
  return s ? (v >> s) | (v << (64 - s)) : v;
}
static inline uint32_t rotl32(uint32_t v, unsigned s) {
  s &= 31u;
  return s ? (v << s) | (v >> (32 - s)) : v;
}
static inline uint32_t rotr32(uint32_t v, unsigned s) {
  s &= 31u;
  return s ? (v >> s) | (v << (32 - s)) : v;
}
static inline int clz64(uint64_t v) { return v ? __builtin_clzll(v) : 64; }
static inline int ctz64(uint64_t v) { return v ? __builtin_ctzll(v) : 64; }

static void test_bits_ops(void) {
  long long *t = alloc_times();
  long long checksum = 0;
  for (int r = 0; r < RUN; r++) {
    long long t0 = now_ns();
    uint64_t acc = 0;
    uint64_t h = 2166136261ULL;
    for (int i = 0; i < 200000; i++) {
      uint64_t x = (uint64_t)i;
      h ^= (x | (x << 3));                 /* bxor(h, bor(x, sl(x,3))) */
      h &= 1152921504606846975ULL;         /* band, 2^60 - 1          */
      h = rotl64(h, 7);                     /* rl64(h,7)               */
      h = rotr64(h, 3);                     /* rr64(h,3)               */
      h ^= (h >> 11);                       /* sr  (logical)           */
      h ^= (uint64_t)((int64_t)h >> 2);     /* sra (arithmetic)        */
      h ^= ~x;                              /* bnot(x)                 */
      acc ^= h;
      acc ^= (uint64_t)rotl32((uint32_t)h, 5);
      acc ^= (uint64_t)rotr32((uint32_t)h, 9);
      acc ^= (uint64_t)__builtin_bswap16((uint16_t)x);
      acc ^= (uint64_t)__builtin_bswap32((uint32_t)x);
      acc ^= __builtin_bswap64(h);
      acc ^= (uint64_t)__builtin_popcountll(h);
      acc ^= (uint64_t)clz64(x);
      acc ^= (uint64_t)ctz64(x);
    }
    checksum = (long long)acc;
    t[r] = now_ns() - t0;
  }
  fprintf(stderr, "bits_ops = %lld\n", checksum);
  record("bits", "ops", t, RUN);
  free(t);
}

void run_bits_group(void) {
  test_bits_ops();
}
